core/src/cnf/mod.rs

```diff
@@ -50,3 +50,12 @@ pub static INSECURE_FORWARD_ACCESS_ERRORS: Lazy<bool> =
 /// If the environment variable is not present or cannot be parsed, a default value of 50,000 is used.
 pub static EXTERNAL_SORTING_BUFFER_LIMIT: Lazy<usize> =
 	lazy_env_parse!("SURREAL_EXTERNAL_SORTING_BUFFER_LIMIT", usize, 50_000);
+
+/// Enable experimental bearer access and stateful access grant management. Still under active development.
+/// Using this experimental feature may introduce risks related to breaking changes and security issues.
+#[cfg(not(test))]
+pub static EXPERIMENTAL_BEARER_ACCESS: Lazy<bool> =
```

Comment:
Instead of an `EXPERIMENTAL_` prefix for this, shouldn't we use a `surrealdb_unstable` flag for this? That way we won't need to break the API unnecessarily to stabilise this.

Reply:
EXPERTIMENTAL makes is clear that you opt into a feature which can change. With `surrealdb_unstable` people would need to compile the db themselves which will limit the amount of people that will test this feature.

Reply:
That was the reasoning, yeah. I proposed a feature flag, an environment flag or an experimental CLI flag and we settled on this as a compromise between allowing users to test the feature without having to compile the binary themselves and unnecessarily exposing them to the risk of breaking changes and security issues that would come by exposing this feature as a CLI flag.

I think breaking changes are to be expected from the `EXPERIMENTAL` prefix and, since this feature will be undocumented, users that become aware of it will do so through either this PR or the code itself, both of which warn about potential breaking changes and security issues while the feature is in experimental state. The [experimental GraphQL feature](https://github.com/surrealdb/surrealdb/pull/3797) is also exposed through an environment variable for similar reasons.

Reply:
On second thought, I am not completely sure on the best approach. May be worth it to discuss further :+1:

Reply:
We have agreed to do both. The feature will be included in the default binaries behind an experimental environment flag but also included under `surrealdb_unstable` so that Rust users are able to opt out of the new unstable code until the feature is stabilized. I will move this PR to draft until this change is done.

Reply:
As we discussed, we will finally use `doc(hidden)` instead of `cfg(surrealdb_unstable)` to prevent breaking changes in the bincode serialization of enums that would include new variants related to this experimental feature. All publicly exposed types related to bearer access have been hidden from the public Rust API both for the `surrealdb` and `surrealdb_core` crates. This ensures that changes can be made without breaking the API contract. This will still include the experimental features in every build, but the code should not be reachable in any case unless the environment flag is activated. For Rust users, the new types will clearly be noted as internal.

Changes made in [554d402](https://github.com/surrealdb/surrealdb/pull/4302/commits/554d402eb9230c9a06279ba5b76c3c52ea86d918).